tools/src/wyvern/target/corewyvernIL/expression/New.java

```diff
@@ -73,37 +85,60 @@ public Declaration findDecl(String name) {
 
 	@Override
 	public ValueType typeCheck(TypeContext ctx) {
-		List<DeclType> dts = new LinkedList<DeclType>();
+List<DeclType> dts = new LinkedList<DeclType>();
 		
 		TypeContext thisCtx = ctx.extend(selfName, getExprType());
 		
-		// check that all decls are well-typed
-		for (Declaration d : decls) {
-			DeclType dt = d.typeCheck(ctx, thisCtx);
-			dts.add(dt);
+		ValueType type = getExprType();
+		if (hasDelegate) {
+			ValueType delegateObjectType = ctx.lookup(delegateDeclaration.getFieldName()); 
+			
+			for (Declaration d : decls_ExceptDelegate()) {
+				DeclType dt = d.typeCheck(ctx, thisCtx);
+				dts.add(dt);
+			}
+			
+			StructuralType delegateStructuralType = delegateObjectType.getStructuralType(thisCtx);
+			// new defined declaration will override delegate object's method definition if they had subType relationship
+			for (DeclType declType : delegateStructuralType.getDeclTypes()) {
+				if (!dts.stream().anyMatch(newDefDeclType-> newDefDeclType.isSubtypeOf(declType, thisCtx))) {
+					dts.add(declType);
+				}
+			}
+		}
+		else {
+			// check that all decls are well-typed
+			for (Declaration d : decls) {
+				DeclType dt = d.typeCheck(ctx, thisCtx);
+				dts.add(dt);
+			}
```

Comment:
This code is present in both the "if" and the "else" blocks, except for using "decls" vs. "decls_ExceptDelegate" (which isn't quite right anyway).  Can you pull it out of the "if" to avoid code duplication?